library/psa_crypto_se.c

```diff
@@ -28,18 +28,33 @@
 #if defined(MBEDTLS_PSA_CRYPTO_SE_C)
 
 #include <assert.h>
+#include <limits.h>
+#include <stdint.h>
 #include <string.h>
 
+#include "psa/crypto_se_driver.h"
+
 #include "psa_crypto_se.h"
 
+#include "mbedtls/platform.h"
+#if !defined(MBEDTLS_PLATFORM_C)
+#define mbedtls_calloc calloc
+#define mbedtls_free   free
+#endif
+
+
+
 /****************************************************************/
 /* Driver lookup */
 /****************************************************************/
 
+typedef struct psa_drv_se_slot_usage_s psa_drv_se_slot_usage_t;
+
 typedef struct psa_se_drv_table_entry_s
 {
     psa_key_lifetime_t lifetime;
     const psa_drv_se_t *methods;
+    psa_drv_se_slot_usage_t *slot_usage;
```

Comment:
Just an observation, looking for a check on my understanding:

It looks like we are currently requiring separate lifetimes for slots that have different capabilities. Is this correct? Is there no other way to correlate what a slot is capable of doing from lifetime?

This seems to require that for raw data import/export slots used for e.g. storing certificates would have a separate lifetime from slots used for ecdsa keys.

Do we also require a new lifetime value for slots that have different hardware-enforced policies? For instance, some slots may be configured as read-only. We don't want to allocate a slot that's read-only for import-key use. We want the slots returned by `psa_drv_cb_find_free_slot()` to be fit for purpose and OK by the desired usage policy.

Reply:
Implementers have some freedom in choosing lifetimes, but the intent is that they characterize the lifetime of a key, not what type of key it is. So slots that can store different types of keys should still have the same lifetime, but erasable and non-erasable slots in the same secure element should have different lifetimes.

Driver can have an `allocate_key` method precisely so that they can pick a slot that's suitable for a given key's parameters (lifetime, type, size, policy, …).

Reply:
Thanks for the explanation. This makes sense now that I understand policy is part of attributes, so that we don't need a new `lifetime` for each policy/usage required.